consumer/src/main/kotlin/streams/kafka/KafkaManualCommitEventConsumer.kt

```diff
@@ -3,15 +3,50 @@ package streams.kafka
 import org.apache.kafka.clients.consumer.OffsetAndMetadata
 import org.apache.kafka.common.TopicPartition
 import org.neo4j.logging.Log
+import streams.extensions.offsetAndMetadata
+import streams.extensions.topicPartition
 import streams.service.StreamsSinkEntity
-import streams.service.errors.ErrorService
+import java.time.Duration
 
 class KafkaManualCommitEventConsumer(config: KafkaSinkConfiguration,
-                                     log: Log): KafkaAutoCommitEventConsumer(config, log) {
+                                     private val log: Log,
+                                     topics: Set<String>): KafkaAutoCommitEventConsumer(config, log, topics) {
+
+    private val asyncCommit = config.streamsAsyncCommit
+
+    override fun stop() {
+        if (asyncCommit) {
+            /*
+             * While everything is fine, we use commitAsync.
+             * It is faster, and if one commit fails, the next commit will serve as a retry.
+             * But if we are closing, there is no "next commit". We call commitSync(),
+             * because it will retry until it succeeds or suffers unrecoverable failure.
+             */
+            consumer.commitSync()
+        }
+        super.stop()
+    }
 
     private fun commitData(commit: Boolean, topicMap: Map<TopicPartition, OffsetAndMetadata>) {
         if (commit) {
-            consumer.commitSync(topicMap)
+            if (asyncCommit) {
+                if (log.isDebugEnabled) {
+                    log.debug("Committing data in async")
+                }
+                consumer.commitAsync(topicMap) { offsets: MutableMap<TopicPartition, OffsetAndMetadata>, exception: Exception? ->
+                    if (exception != null) {
+                        log.warn("""
+                            |These offsets `$offsets`
```

Comment:
message should maybe mention topic name; user may have many topics configured, and offsets are just integers, right?

Reply:
`offsets` is a map of `<TopicPartition, OffsetAndMetadata>` so we print the information about:
- topic
- partition
- offset
- metadata